src/urllib3/response.py

```diff
@@ -869,6 +869,12 @@ def _raw_read(
                     # raised during streaming, so all calls with incorrect
                     # Content-Length are caught.
                     raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
+            elif read1 and (not data or self.length_remaining == len(data)):
```

Comment:
`read1` can return zero bytes without the stream being finished, the semantics of `read1` are different than `read`: `read1(n)` returns up to N bytes, `read(n)` returns either exactly N bytes or if the stream is complete less than N bytes.

For this reason we can't have the `not data` case, otherwise we're at risk of closing our HTTP response early.

Reply:
Docs [state](https://docs.python.org/3/library/io.html#io.BufferedIOBase.read1) "If size is -1 (the default), an arbitrary number of bytes are returned (**more than zero unless EOF is reached**)." It looks like [CPython's source](https://github.com/python/cpython/blob/6a1d5a4c0f209d51ab33d6529935d643bcdb3ba2/Lib/http/client.py#L664-L665) expects the same "more than zero" for sizes greater than 0 too. Am I missing something?

Your comment made me think about the case when user calls `.read1(0)`, we shouldn't close the IO in such a case.

Reply:
Hmm, maybe I misunderstood the contract of `read1`, so it's identical to `read` only without the chaining, cool. Thanks Illia!